rfc/src/rfcs/0002-function-stubbing.md

```diff
@@ -0,0 +1,387 @@
+Furthermore, contract predicates constitute a DSL of their own that needs to be learned; using stubbing, we can stick to using just Rust.
+
+### Alternative #1: Annotate stubbed functions
+
+In this alternative, users add an attribute `#[kani::stub_by(<replacement>)]` to the function that should be replaced.
+This approach is similar to annotating a function with a contract specifying its behavior (the stub acts like a programmatic contract).
+The major downside with this approach is that it would not be possible to stub external code. We see this as a likely use case that needs to be supported: users will want to replace `std` library functions or functions from arbitrary external crates.
+
+### Alternative #2: Annotate stubs
+
+In this alternative, users add an attribute `#[kani::stub(<original>)]` to the stub function itself, saying which function it replaces:
+
+```rust
+#[cfg(kani)]
+#[kani::stub("rand::random")]
+fn mock_random<T: kani::Arbitrary>() -> T { ... }
+```
+
+The downside is that this stub must be uniformly applied across all harnesses and the stub specifications might be spread out across multiple files.
+It would also require an extra layer of indirection to use a function as a stub if the user does not have source code access to it.
+
+### Alternative #3: Annotate harnesses and stubs 
+
+This alternative combines the proposed solution and Alternative #2.
+Users annotate the stub (as in Alternative #2) and specify for each harness which stubs to use using an annotation `#[kani::use_stubs(<stub>+)]` placed above the harness.
+
+This could be combined with modules, so that a module can be used to group stubs together, and then harnesses could pull in all the stubs in the module:
+
+```rust
+#[cfg(kani)]
+mod my_stubs {
+
+  #[kani::stub("foo")]
+  fn stub1() { ... }
+
+  #[kani::stub("bar")]
+  fn stub2() { ... }
+
+}
+
+#(cfg[kani])
+#[kani::proof]
+#[kani::use_stubs("my_stubs")]
+fn my_harness() { ... }
+```
+
+The benefit is that stubs are specified per harness, and (using modules) it might be possible to group stubs together.
+The downside is that multiple annotations are required and the stub mappings themselves are remote from the harness.
+
+### Alternative #4: Specify stubs in a file 
+
+One alternative would be to specify stubs in a file that is passed to `kani-driver` via a command line option.
+Users would specify per-harness stub pairings in the file; JSON would be a possible format.
+Using a file would eliminate the need for `kani-driver` to extract harness information from the Rust source code; the rest of the implementation would stay the same.
+It would also allow the same harness to be run with different stub selections (by supplying a different file).
+The disadvantage is that the stub selection is remote from the harness itself.
+
+## Rationale and alternatives: stubbing mechanism
+
+Our approach is based on a MIR-to-MIR transformation.
```

Comment:
I would highlight that another benefit is that we can extend our compiler to integrate `--concrete-playback` with the abstractions. But that this is out of the scope for this work.